blog post draft "Fixing a bug in PyPy's incremental GC" (#112)
pypy/pypy.org: posts/2024/03/fixing-bug-incremental-gc.md

.. tags: 
.. category: 
.. link: 
.. description: 
.. type: text
.. author: Carl Friedrich Bolz-Tereick
-->

# Introduction

Since last summer, I've been looking on and off into a weird and hard to
reproduce [crash bug in PyPy](https://github.com/pypy/pypy/issues/3959). It was
manifesting only on CI, and it seemed to always happening in the AST rewriting
phase of [pytest](https://pytest.org), the symptoms being that PyPy would crash
with a segfault. All my attempts to reproduce it locally failed, and my
attempts to try to understand the problem by dumping the involved ASTs lead
nowhere.

A few weeks ago, we got [two more](https://github.com/PyO3/pyo3/issues/3766)
[bug reports](https://github.com/orgs/pypy/discussions/4923), the last one by
the authors of the [nanobind](https://nanobind.readthedocs.io/) binding
generator, with the same symptoms: crash in AST rewriting, only on CI. I
decided to make a more serious push to try to find the bug this time.
Ultimately the problem turned out to be several bugs in PyPy's garbage
collector (GC) that had been there since its inception in
[2013](https://www.pypy.org/posts/2013/10/incremental-garbage-collector-in-pypy-8956893523842234676.html)

> (started by [Andrew Chambers](https://acha.ninja/)). Understanding the

Reviewer: It sounds a little like you are blaming Andrew. Maybe leave this part out?
```suggestion
Understanding the
```

situation turned out to be quite involved, additionally complicated by this
being the first time that I was working on this particular aspect of PyPy's GC.
Since the bug was so much work to find, I thought I'd write a blog post about
it.

The blog post consists of three parts: first a chronological description of
what I did to find the bug, a technical explanation of what goes wrong, some
reflections on the bug (and then a bonus bug I also found in the process).

# Finding the Bug

> I was starting from the failing [nanobind

Reviewer: ```suggestion
I started from the failing [nanobind
```

CI](https://github.com/wjakob/nanobind/actions/runs/8234561874/job/22516568891)
runs that ended with a segfault of the PyPy interpreter. This was only an
intermittent problem, not every run was failing. When I tried to just run the
test suite locally, I couldn't get it to fail. Therefore at first I tried to
learn more about what was happening by looking on the CI runners.

## Running on CI

I forked the nanobind repo and hacked the CI script in order to get it to use a

> PyPy build with full debug information and more assertions turned on. In order

Reviewer: Is there a link how to do this?
```suggestion
PyPy build with [full debug information and more assertions turned on](https://doc.pypy.org/en/latest/build.html#making-a-debug-build-of-pypy). In order
```

variable to the CI script that just contained 32 parameters. This means every
build is done 32 times (sorry Github for wasting your CPUs 😕). With that
amount of repetition, one of the builds was quite reliably crashing.

Then I added the `-Xfaulthandler` option to the PyPy command which will use the
[faulthandler](https://docs.python.org/3.11/library/faulthandler.html) module
try to print a Python stacktrace if the VM segfaults to confirm that PyPy was
indeed crashing in the [AST](https://docs.python.org/3/library/ast.html)
[rewriting
phase](https://github.com/pytest-dev/pytest/blob/main/src/_pytest/assertion/rewrite.py)
of pytest, which pytest uses for [nicer
assertions](https://docs.pytest.org/en/7.1.x/how-to/assert.html#asserting-with-the-assert-statement).
I experimented with hacking our faulthandler implementation to also give me a
C-level callstack, but that didn't work as well as I hoped.

Then I tried to run [gdb](https://sourceware.org/gdb/) on CI to try to get it
to print a C callstack at the crash point. You can get gdb to execute commands
as if typed at the prompt with the `-ex` commandline option, I used something
like this:

```

> gdb -ex "set confirm off" -ex "set pagination off" -ex "set debuginfod enabled off" -ex run -ex where -ex quit --args <command> <arguments>

Reviewer: Will this render more nicely if it is wrapped?
```suggestion
gdb -ex "set confirm off" -ex "set pagination off" -ex \
    "set debuginfod enabled off" -ex run -ex where -ex quit \
    --args <command> <arguments>
```

Author: yeah, it should, I'll try it.

Extremely frustratingly, after two days of work I concluded that the assertion
logic itself was wrong. I have fixed that in the meantime too, the gory details
of that are in the bonus section at the end of the post.

## Using GDB scripting to find the real bug

After that disaster I went back to the earlier rr recording without assertions
and tried to understand in more detail why the GC decided to free an object
that was still being referenced. To be able to do that I used the [GDB Python
scripting
API](https://sourceware.org/gdb/current/onlinedocs/gdb.html/Python-API.html) to
write some helper commands to understand the state of the GC heap (rr is an
extension of GDB, so the GDB scripting API works in rr too).

The first (small) helper command I wrote with the GDB scripting API was a way
to pretty-print the currently active GC flags of a random PyPy object, starting
just from the pointer. The more complex command I wrote was an object tracer,
which follows pointers to GC objects starting from a root object to explore the
object graph. The object tracer isn't complete, it doesn't deal with all the
complexities of PyPy's GC. But it was good enough to help me with my problem, I

> found out that the corrupted object was stored in an array.

Reviewer: Is there a way to preserve the helpers in a wiki or as part of the pypy repo?

Author: good question! I'll look into it

Author: I've added a gist link to the full code

	node = self.malloc(S) # unrelated object, will be collected
	node.x = 5
	self.writearray(source, 0, node) # store reference into source array, calling the write barrier
	val = self.gc.collect_step() #
	source = self.stackroots[0] # reload arrays, they might have move
	target = self.stackroots[1]
	# this GC step traces target
	val = self.gc.collect_step()

	# emulate what a memcopy of arrays does
	addr_src = llmemory.cast_ptr_to_adr(source)
	addr_dst = llmemory.cast_ptr_to_adr(target)
	res = self.gc.writebarrier_before_copy(addr_src, addr_dst, 0, 0, 2)
	assert res
	target[0] = source[0] # copy two elements of the arrays
	target[1] = source[1]
	self.writearray(source, 0, lltype.nullptr(S))
	# this GC step traces source
	self.gc.collect_step()
	# some more collection steps, crucially target isn't traced again
	# but node is deleted
	for i in range(3):
		self.gc.collect_step()
	# used to crash, node got collected
	assert target[0].x == 5

```

One of the fun properties of testing our GC that way is that all the memory is
emulated. The crash in the last line of the test isn't a segfault at all,
instead you get a nice exception saying that you tried to access a freed chunk

> of memory and you can then debug this with a python2 debugger.

Reviewer: I like the way you (subtly?) (not so subtly?) drive home the point that writing in RPython has advantages for dense complicated code like GC implementations.

Author: I think it is actually an advantage, yes. the fact that we have debugged a bunch of GC problems in GDB is kind of a sign that we under-tested our rpython code as python2 code.

and major collections (collecting long-lived objects eventually, using a
mark-sweep algorithm). Young objects are allocated in a nursery using a
bump-pointer allocator, which makes allocation quite efficient. They are moved
out of the nursery by minor collections. In order to find references from old
to young objects the GC uses a write barrier to detect writes into old objects.

The GC is also
[incremental](https://en.wikipedia.org/wiki/Tracing_garbage_collection#Stop-the-world_vs._incremental_vs._concurrent),
which means that its major collections aren't done all at once (which would
lead to long pauses). Instead, major collections are sliced up into small
steps, which are done directly after a minor collection (the GC isn't
*concurrent* though, which would mean that the GC does work in a separate
thread).

The incremental GC uses [tri-color
marking](https://en.wikipedia.org/wiki/Tracing_garbage_collection#Tri-color_marking)
to reason about the reachable part of the heap during the marking phase, where
old objects can be:

- black: already marked, reachable, definitely survives the collection
- grey: will survive, but still needs to be marked

> - white: potentially dead The color of every object is encoded by setting flags

Reviewer: ```suggestion
- white: potentially dead 

The color of every object is encoded by setting flags
```

  in the object header.

The GC maintains the **invariant** that black objects must never point to white
objects. At the start of a major collection cycle the stack roots are turned
gray. During a major collection cycle, the GC will trace gray objects, until
none are left. To trace a gray object, all the objects it references have to be
marked grey if they are white so far. After a grey object is traced, it can be
marked black (because all the referenced objects are now either black or gray).
Eventually, there are no gray objects left. At that point (because no white
object can be reached from a black one) all the white objects are known to be
unreachable and can therefore be freed.

The GC is incremental because every collection step will only trace a limited

> amount of gray object, before giving control back to the program. This leads to

Reviewer: ```suggestion
number of gray objects, before giving control back to the program. This leads to
```

> a problem: if an already traced (black) object is changed between to marking

Reviewer: ```suggestion
a problem: if an already traced (black) object is changed between two marking
```

steps of the GC, the program can mutate that object and write a new reference
into one of its field. This could lead to an invariant violation, if the
referenced object is white. Therefore, the GC uses the write barrier (which it

> needs anyway to find references from old to young objects) to mark all black
> objects that are written into gray, and then trace them again at one of the

Reviewer: Not clear what "written" is: "traced" or "modified".
```suggestion
needs anyway to find references from old to young objects) to mark all black
objects that are written into gray, and then trace them again at one of the
```

Author: I replaved "written into" with "modified" here.

later collection steps.

## The special write barrier of memcopy

Arrays use a different kind of write barrier than normal objects. Since they
can be arbitrarily large, tracing them can take a long time. Therefore it's
potentially wasteful to trace them fully at a minor collection. To fix this.
the array write barrier keeps more granular information about which parts of

> the array have been written to since the last collection step. Then only the

Reviewer: ```suggestion
the array have been modified since the last collection step. Then only the
```

during a marking phase. This was simply not done at all for the memcopy write
barrier, at least in two of the code paths. Fixing this problem fixes the unit
tests and stops the crashes.

# Reflections

The way the bug was introduced is really typical. A piece of code (the memcopy
write barrier) was written under a set of assumptions. Then those assumptions
changed later. Not all the code pieces that relied on these assumptions to be
correct were updated. It's pretty hard to prevent this in all situations.

I still think we could have done more to prevent the bug occurring. Writing a
property-based test for the GC would have been a good idea given the complexity
of the GC, and definitely something we did in other parts of our code at the
time (just using the `random` module mostly, we started using hypothesis
later).

It's a bit of a mystery to me why this bug managed to be undetected for so
long. Memcopy happens in a lot of pretty core operations of e.g. lists in
Python (`list.extend`, to name just one example). To speculate, I would suspect
that all the other preconditions for the bug occurring made it pretty rare:

> - the contend of an old list that is not yet marked needs to be copied into

Reviewer: ```suggestion
- the content of an old list that is not yet marked needs to be copied into
```

  another old list that is marked already 
- the source of the copy needs to also store an object that has no other
  references 
- the source of the copy then needs to be overwritten with other data 
- then the next collection steps need to be happening at the right points
- ...

Given the complexity of the GC logic I also wonder whether some lightweight
formal methods would have been a good idea. Formalizing some of the core
invariants in B or TLA+ and then model checking them up to some number of

> objects would have found this problem pretty soon. There are also correctness

Reviewer: ```suggestion
objects would have found this problem pretty quickly. There are also correctness
```

PyPy's VM-building bootstrapping process can "freeze" a bunch of heap objects
into the final binary. This allows the VM to start up quickly, because those
frozen objects are loaded by the OS as part of the binary.

Those frozen pre-built objects are parts of the 'roots' of the garbage
collector and need to be traced. However, tracing all the pre-built objects at
every collection would be very expensive, because there are a lot of them
(about 150,000 in a PyPy 3.10 binary). Tracing them all is also not necessary,
because most of them are never modified. Therefore they only have references to
other pre-built objects, which can never be deallocated anyway. Therefore we
have an optimization that uses the write barrier (which we need anyway to find
old-to-young pointers) to notice when a pre-built object gets modified for the
very first time. If that happens, it gets added to the set of pre-built objects
that gets counted as a root, and is therefore traced as a root at collections
from then on.

## The wrong assertion

The assertion that triggered when I turned on the GC debug mode was saying that
the GC found a reference from a black to a white object, violating its

> invariant. pre-built objects count as always black, as long as they have never

Reviewer: ```suggestion
invariant. Pre-built objects count as always black, as long as they have never
```

Author: wow, this sentence is pretty convoluted, I think I'll try to reformulate it

Author: done

.. slug: fixing-bug-incremental-gc
.. date: 2024-03-26 19:14:09 UTC
.. tags:
.. category:
.. link:
.. description:
.. type: text
.. author: Carl Friedrich Bolz-Tereick
-->

# Introduction

Since last summer, I've been looking on and off into a weird and hard to
reproduce [crash bug in PyPy](https://github.com/pypy/pypy/issues/3959). It was
manifesting only on CI, and it seemed to always happen in the AST rewriting
phase of [pytest](https://pytest.org), the symptoms being that PyPy would crash
with a segfault. All my attempts to reproduce it locally failed, and my
attempts to try to understand the problem by dumping the involved ASTs lead
nowhere.

A few weeks ago, we got [two more](https://github.com/PyO3/pyo3/issues/3766)
[bug reports](https://github.com/orgs/pypy/discussions/4923), the last one by
the authors of the [nanobind](https://nanobind.readthedocs.io/) binding
generator, with the same symptoms: crash in AST rewriting, only on CI. I
decided to make a more serious push to try to find the bug this time.
Ultimately the problem turned out to be several bugs in PyPy's garbage
collector (GC) that had been there since its inception in

> [2013](https://www.pypy.org/posts/2013/10/incremental-garbage-collector-in-pypy-8956893523842234676.html).

Author: I wonder whether nikola is configured to do something special for links that go to the same domain. locally at least replacing it with a relative link fixes it, will try to push

Understanding the
situation turned out to be quite involved, additionally complicated by this
being the first time that I was working on this particular aspect of PyPy's GC.
Since the bug was so much work to find, I thought I'd write a blog post about
it.

The blog post consists of three parts: first a chronological description of
what I did to find the bug, a technical explanation of what goes wrong, some
reflections on the bug (and then a bonus bug I also found in the process).

# Finding the Bug

I started from the failing [nanobind CI

> runs](https://github.com/wjakob/nanobind/actions/runs/8234561874/job/22516568891)

Reviewer: This link will disappear in a few months...

Author: I think I don't care too much. we link to the discussion too, that's good enough when the action run gets removed

<script src="https://asciinema.org/a/648877.js" id="asciicast-648877" async="true"></script>

([Max](https://bernsteinbear.com/) pointed out
[autoclave](https://github.com/silentbicycle/autoclave) to me when reviewing
this post, which is a more dedicated tool for this job.)

Thankfully, running the tests repeatedly eventually lead to a crash, solving my
"only happens on CI" problem. I then tried various variants to exclude possible
sources of errors. The first source of errors to exclude in PyPy bugs is the
just-in-time compiler, so I reran the tests with `--jit off` to see whether I
could still get it to crash, and thankfully I eventually could (JIT bugs are
often very annoying).

Next source of bugs to exclude where C-extensions. Since those were the tests
of nanobind, a framework for creating C-extension modules I was a bit worried
that the bug might be in our emulation of CPython's C-API. But running PyPy
with the `-v` option (which will print all the imports as they happen)
confirmed that at the point of crash no C-extension had been imported yet.

## Using `rr`

I still couldn't get the bug to happen in GDB, so the tool I tried next was

> [rr](https://rr-project.org/). rr can record the execution of a program and

Reviewer: the link gets a bit lost, maybe something like this? Or am I getting too carried away :)
```suggestion
[rr, the "reverse debugger"](https://rr-project.org/). rr can record the execution of a program and
```

later replay it arbitrarily often. This gives you a time-traveling debugger
that allows you to execute the program backwards in addition to forwards.

> Eventually I managed to get the crash to happen when running the tests with `rr
> record --chaos` (`--chaos` randomizes some decisions that rr takes, to try to

Reviewer: Line break in the code fragment renders as a line break, so avoid that
```suggestion
Eventually I managed to get the crash to happen when running the tests with
`rr record --chaos` (`--chaos` randomizes some decisions that rr takes, to try to
```

into the first word of the object, overwriting the object's header. The next
time the object was used things crashed.

## Side-quest: wrong GC assertions

At this point in the process, I got massively side-tracked. PyPy's GC has a
number of debug modes that you can optionally turn on. Those slow down the
program execution a lot, but they should in theory help to understand why the
GC goes wrong. When I turned them on, I was getting a failing assertion really
early in the test execution, complaining about an invariant violation in the GC
logic. At first this made me very happy. I thought that this would help me fix
the bug more quickly.

Extremely frustratingly, after two days of work I concluded that the assertion
logic itself was wrong. I have fixed that in the meantime too, the details
of that are in the bonus section at the end of the post.

## Using GDB scripting to find the real bug

After that disaster I went back to the earlier rr recording without GC assertions
and tried to understand in more detail why the GC decided to free an object
that was still being referenced. To be able to do that I used the [GDB Python
scripting

> API](https://sourceware.org/gdb/current/onlinedocs/gdb.html/Python-API.html) to

Reviewer: This link renders well, so the problem above is not connected to having an `.html` suffix.

testing](https://hypothesis.readthedocs.io/en/latest/stateful.html) approach of
hypothesis, but I just implemented it manually with the [data
strategy](https://hypothesis.readthedocs.io/en/latest/data.html#drawing-interactively-in-tests).

Every one of those steps is always performed on both the tested GC, and on some
regular Python objects. The Python objects provide the "ground truth" of what
the heap should look like, so we can compare the state of the GC objects
with the state of the Python objects to find out whether the GC made a mistake.

In order to check whether the test is actually useful, I reverted my bug fixes
and made sure that the test re-finds both the spurious GC assertion error and the
problems with memcopying an array.

In addition, the test also found corner cases in my fix. There was a situation
that I hadn't accounted for, which the test found after eventually.
I also plan on adding a bunch of other GC features as steps in the
test to stress them too (for example weakrefs, identity hashes, pinning, maybe
finalization).

At the point of publishing this post, the fixes got merged to the 2.7/3.9/3.10

> branches of PyPy, and we'll have to do a release with the fixes at some point.

Reviewer: ```suggestion
branches of PyPy, and will be part of the next release (v7.3.16).
```

In addition, there is another optimization for arrays, which is that memcopy is
treated specially by the GC. If memcopy is implemented by simply writing a loop
that copies the content of one array to the other, that will invoke the write
barrier every single loop iteration for the write of every array element,
costing a lot of overhead. Here's some pseudo-code:

```python
def arraycopy(source, dest, source_start, dest_start, length):
    for i in range(length):
        value = source[source_start + i]
        dest[dest_start + i] = value # <- write barrier inserted here
```

Therefore the GC has a special memcopy-specific
write barrier that will perform the GC logic once before the memcopy loop, and
then use a regular (typically SIMD-optimized) memcopy implementation from
`libc`. Roughly like this:

```python
def arraycopy(source, dest, source_start, dest_start, length):
    gc_writebarrier_before_array_copy(source, dest, source_start, dest_start, length)
    raw_memcopy(cast_to_voidp(source) + source_start,
                cast_to_voidp(dest) + dest_start,
                sizeof(itemtype(source)) * length)
```

(this is really a rough sketch. The [real

> code](ttps://github.com/pypy/pypy/blob/789f964fff59c722b0872abcdc56d2b1373a9f3b/rpython/rlib/rgc.py#L365)

Reviewer: ```suggestion
code](https://github.com/pypy/pypy/blob/789f964fff59c722b0872abcdc56d2b1373a9f3b/rpython/rlib/rgc.py#L365)
```
